From 0c2fae4d73834cfbe4eacb88736443121102838d Mon Sep 17 00:00:00 2001
From: Henry Wang <Henry.Wang@arm.com>
Date: Sat, 8 Oct 2022 07:11:19 +0000
Subject: [PATCH] tools/xl-psr: Support SLC in `xl psr-hwinfo` cmd

Users can leverage the `xl psr-hwinfo` command to get the hardware
information of resources that are used by PSR/MPAM to control.
Currently, this command only supports the L2, L3 and memory bandwidth.

Extend the functionality of `xl psr-hwinfo` to the detection of
system level cache (SLC) on Arm if SLC is available and controlled by
MPAM, and use level 0 in the command to represent the SLC. To implement
that, interfaces `{libxl,xc}_psr_cmt_get_slc_size()` are introduced to
libxl and libxc respectively, considering changing original interfaces
`{libxl,xc}_psr_cmt_get_l3_cache_size()` might introduce a breakage of
API. These newly introduced interfaces take the advantage of sysctl
subops added by the previous patch.

Signed-off-by: Henry Wang <Henry.Wang@arm.com>
Upstream-Status: Pending
Issue-Id: SCM-5244
Change-Id: Ic8612cfa2d6623d007cd28ce3b3340638a0034be
---
 tools/include/libxl.h        |  2 ++
 tools/include/xenctrl.h      |  3 +++
 tools/libs/ctrl/xc_psr.c     | 40 ++++++++++++++++++++++++++++++------
 tools/libs/light/libxl_psr.c | 29 +++++++++++++++++++++++++-
 tools/xl/xl_psr.c            | 20 +++++++++++++++++-
 xen/include/public/sysctl.h  |  4 ++++
 6 files changed, 90 insertions(+), 8 deletions(-)

diff --git a/tools/include/libxl.h b/tools/include/libxl.h
index 5d5b57f842..f9b4f480b5 100644
--- a/tools/include/libxl.h
+++ b/tools/include/libxl.h
@@ -2737,6 +2737,8 @@ int libxl_psr_cmt_get_total_rmid(libxl_ctx *ctx, uint32_t *total_rmid);
 int libxl_psr_cmt_get_l3_cache_size(libxl_ctx *ctx,
                                     uint32_t socketid,
                                     uint32_t *l3_cache_size);
+int libxl_psr_cmt_get_slc_size(libxl_ctx *ctx, uint32_t socketid,
+                               uint32_t *slc_size);
 int libxl_psr_cmt_get_cache_occupancy(libxl_ctx *ctx,
                                       uint32_t domid,
                                       uint32_t socketid,
diff --git a/tools/include/xenctrl.h b/tools/include/xenctrl.h
index c009946f18..660c2321e0 100644
--- a/tools/include/xenctrl.h
+++ b/tools/include/xenctrl.h
@@ -2493,6 +2493,7 @@ typedef enum xc_psr_type xc_psr_type;
 enum xc_psr_feat_type {
     XC_PSR_CAT_L3,
     XC_PSR_CAT_L2,
+    XC_PSR_CAT_SLC,
     XC_PSR_MBA,
 };
 typedef enum xc_psr_feat_type xc_psr_feat_type;
@@ -2522,6 +2523,8 @@ int xc_psr_cmt_get_l3_upscaling_factor(xc_interface *xch,
 int xc_psr_cmt_get_l3_event_mask(xc_interface *xch, uint32_t *event_mask);
 int xc_psr_cmt_get_l3_cache_size(xc_interface *xch, uint32_t cpu,
                                  uint32_t *l3_cache_size);
+int xc_psr_cmt_get_slc_size(xc_interface *xch, uint32_t cpu,
+                            uint32_t *slc_size);
 int xc_psr_cmt_get_data(xc_interface *xch, uint32_t rmid, uint32_t cpu,
                         xc_psr_cmt_type type, uint64_t *monitor_data,
                         uint64_t *tsc);
diff --git a/tools/libs/ctrl/xc_psr.c b/tools/libs/ctrl/xc_psr.c
index 1a0ab636f0..0e834fbb5d 100644
--- a/tools/libs/ctrl/xc_psr.c
+++ b/tools/libs/ctrl/xc_psr.c
@@ -158,6 +158,30 @@ int xc_psr_cmt_get_l3_cache_size(xc_interface *xch, uint32_t cpu,
     return rc;
 }
 
+int xc_psr_cmt_get_slc_size(xc_interface *xch, uint32_t cpu, uint32_t *slc_size)
+{
+    static int val = 0;
+    int rc;
+    DECLARE_SYSCTL;
+
+    if ( val )
+    {
+        *slc_size = val;
+        return 0;
+    }
+
+    sysctl.cmd = XEN_SYSCTL_psr_cmt_op;
+    sysctl.u.psr_cmt_op.cmd = XEN_SYSCTL_PSR_CMT_get_slc_size;
+    sysctl.u.psr_cmt_op.flags = 0;
+    sysctl.u.psr_cmt_op.u.slc.cpu = cpu;
+
+    rc = xc_sysctl(xch, &sysctl);
+    if ( !rc )
+        val = *slc_size= sysctl.u.psr_cmt_op.u.data;
+
+    return rc;
+}
+
 int xc_psr_cmt_get_data(xc_interface *xch, uint32_t rmid, uint32_t cpu,
                         xc_psr_cmt_type type, uint64_t *monitor_data,
                         uint64_t *tsc)
@@ -346,11 +370,15 @@ int xc_psr_get_hw_info(xc_interface *xch, uint32_t socket,
 
     switch ( type )
     {
+    case XC_PSR_CAT_SLC:
     case XC_PSR_CAT_L2:
     case XC_PSR_CAT_L3:
-        sysctl.u.psr_alloc.cmd = (type == XC_PSR_CAT_L2) ?
-                                 XEN_SYSCTL_PSR_get_l2_info :
-                                 XEN_SYSCTL_PSR_get_l3_info;
+        if ( type == XC_PSR_CAT_SLC )
+            sysctl.u.psr_alloc.cmd = XEN_SYSCTL_PSR_get_slc_info;
+        if ( type == XC_PSR_CAT_L2 )
+            sysctl.u.psr_alloc.cmd = XEN_SYSCTL_PSR_get_l2_info;
+        if ( type == XC_PSR_CAT_L3 )
+            sysctl.u.psr_alloc.cmd = XEN_SYSCTL_PSR_get_l3_info;
 
         rc = xc_sysctl(xch, &sysctl);
         if ( rc )
@@ -358,10 +386,10 @@ int xc_psr_get_hw_info(xc_interface *xch, uint32_t socket,
 
         hw_info->cat.cos_max = sysctl.u.psr_alloc.u.cat_info.cos_max;
         hw_info->cat.cbm_len = sysctl.u.psr_alloc.u.cat_info.cbm_len;
-        hw_info->cat.cdp_enabled = (type == XC_PSR_CAT_L2) ?
-                                   false :
+        hw_info->cat.cdp_enabled = (type == XC_PSR_CAT_L3) ?
                                    (sysctl.u.psr_alloc.u.cat_info.flags &
-                                    XEN_SYSCTL_PSR_CAT_L3_CDP);
+                                    XEN_SYSCTL_PSR_CAT_L3_CDP) :
+                                   false;
 
         break;
     case XC_PSR_MBA:
diff --git a/tools/libs/light/libxl_psr.c b/tools/libs/light/libxl_psr.c
index 9ced7d1715..4a4a67c715 100644
--- a/tools/libs/light/libxl_psr.c
+++ b/tools/libs/light/libxl_psr.c
@@ -231,6 +231,31 @@ out:
     return rc;
 }
 
+int libxl_psr_cmt_get_slc_size(libxl_ctx *ctx, uint32_t socketid,
+                               uint32_t *slc_size)
+{
+    GC_INIT(ctx);
+
+    int rc;
+    int cpu = libxl__pick_socket_cpu(gc, socketid);
+
+    if (cpu < 0) {
+        LOGE(ERROR, "failed to get socket cpu");
+        rc = ERROR_FAIL;
+        goto out;
+    }
+
+    rc = xc_psr_cmt_get_slc_size(ctx->xch, cpu, slc_size);
+    if (rc < 0) {
+        libxl__psr_cmt_log_err_msg(gc, errno);
+        rc = ERROR_FAIL;
+    }
+
+out:
+    GC_FREE;
+    return rc;
+}
+
 int libxl_psr_cmt_type_supported(libxl_ctx *ctx, libxl_psr_cmt_type type)
 {
     GC_INIT(ctx);
@@ -345,12 +370,14 @@ static xc_psr_feat_type libxl__feat_type_to_libxc_feat_type(
 
     switch (type) {
     case LIBXL_PSR_FEAT_TYPE_CAT:
-        assert(lvl == 3 || lvl == 2);
+        assert(lvl == 3 || lvl == 2 || lvl == 0);
 
         if (lvl == 3)
             xc_type = XC_PSR_CAT_L3;
         if (lvl == 2)
             xc_type = XC_PSR_CAT_L2;
+        if (lvl == 0)
+            xc_type = XC_PSR_CAT_SLC;
         break;
     case LIBXL_PSR_FEAT_TYPE_MBA:
         xc_type = XC_PSR_MBA;
diff --git a/tools/xl/xl_psr.c b/tools/xl/xl_psr.c
index 38750a149a..4b7cd69f91 100644
--- a/tools/xl/xl_psr.c
+++ b/tools/xl/xl_psr.c
@@ -34,6 +34,7 @@
 #endif
 
 enum psr_cat_type {
+    PSR_CAT_SLC,
     PSR_CAT_L2 = 2,
     PSR_CAT_L3,
 };
@@ -309,12 +310,15 @@ static int psr_cat_hwinfo(enum psr_cat_type type)
 {
     int rc;
     unsigned int i, nr;
-    uint32_t l3_cache_size;
+    uint32_t l3_cache_size, slc_size;
     libxl_psr_cat_info *info;
     char *cache_type;
 
     switch (type)
     {
+    case PSR_CAT_SLC:
+        cache_type = " SLC";
+        break;
     case PSR_CAT_L2:
         cache_type = " L2";
         break;
@@ -341,6 +345,14 @@ static int psr_cat_hwinfo(enum psr_cat_type type)
                 goto out;
             }
         }
+        else if (type == PSR_CAT_SLC) {
+            rc = libxl_psr_cmt_get_slc_size(ctx, info[i].id, &slc_size);
+            if (rc) {
+                fprintf(stderr, "Failed to get SLC size for socket:%d\n",
+                        info[i].id);
+                goto out;
+            }
+        }
 
         printf("%-16s: %u\n", "Socket ID", info[i].id);
 
@@ -350,6 +362,8 @@ static int psr_cat_hwinfo(enum psr_cat_type type)
             printf("%-16s: %s\n", "CDP Status",
                    info[i].cdp_enabled ? "Enabled" : "Disabled");
         }
+        else if (type == PSR_CAT_SLC)
+            printf("%-16s: %uKB\n", "SLC", slc_size);
 
         printf("%-16s: %u\n", "Maximum "PARTITION_IDENTIFIER_NAME,
                info[i].cos_max);
@@ -794,6 +808,10 @@ int main_psr_hwinfo(int argc, char **argv)
     if (all || cat)
         ret = psr_cat_hwinfo(PSR_CAT_L2);
 
+    /* SLC CAT is independent of CMT and L2/L3 CAT, and only available on Arm */
+    if (all || cat)
+        ret = psr_cat_hwinfo(PSR_CAT_SLC);
+
     /* MBA is independent of CMT and CAT */
     if (all || mba)
         ret = psr_mba_hwinfo();
diff --git a/xen/include/public/sysctl.h b/xen/include/public/sysctl.h
index c863adc42c..270a5754f7 100644
--- a/xen/include/public/sysctl.h
+++ b/xen/include/public/sysctl.h
@@ -714,6 +714,10 @@ struct xen_sysctl_psr_cmt_op {
             uint32_t cpu;   /* IN */
             uint32_t rsvd;
         } l3_cache;
+        struct {
+            uint32_t cpu;   /* IN */
+            uint32_t rsvd;
+        } slc;
     } u;
 };
 
-- 
2.25.1

